### PR TITLE
virt: use XENFEAT_dom0 to detect the hardware domain (#6442, #6662)

### PR DESCRIPTION
**Pull request by olafhering**
The detection of ConditionVirtualisation= relies on the presence of
/proc/xen/capabilities. If the file exists and contains the string
"control_d", the running system is a dom0 and VIRTUALIZATION_NONE should
be set. In case /proc/xen exists, or some sysfs files indicate "xen",
VIRTUALIZATION_XEN should be set to indicate the system is a domU.

With an (old) xenlinux based kernel, /proc/xen/capabilities is always
available and the detection described above works always. But with a
pvops based kernel, xenfs must be mounted on /proc/xen to get
"capabilities". This is done by a proc-xen.mount unit, which is part of
xen.git. Since the mounting happens "late", other units may be scheduled
before "proc-xen.mount". If these other units make use of
"ConditionVirtualisation=", the virtualization detection returns
incorect results. detect_vm() will set VIRTUALIZATION_XEN because "xen"
is found in sysfs. This value will be cached. Once xenfs is mounted, the
next process that runs detect_vm() will get VIRTUALIZATION_NONE.

This misdetection can be fixed by using
/sys/hypervisor/properties/features, which exports the value returned by
the "XENVER_get_features" hypercall. If the bit XENFEAT_dom0 is set, the
domain is the "hardware domain". It is supposed to have permissions to
access all hardware. The used sysfs file is available since v2.6.31.

The commonly used term "dom0" refers to the control domain which runs
the toolstack and has access to all hardware. But the virtualization
host may be configured such that one dedicated domain becomes the
"hardware domain", and another one the "toolstack domain".

### CODE REVIEW

File: src/basic/virt.c [hunk 1]
```diff
@@ -227,12 +227,31 @@ static int detect_vm_xen(void) {
         return VIRTUALIZATION_XEN;
 }
 
+#define XENFEAT_dom0 11 /* xen/include/public/features.h */
 /* Returns -errno, or 0 for domU, or 1 for dom0 */
 static int detect_vm_xen_dom0(void) {
         _cleanup_free_ char *domcap = NULL;
+        static const char path_features[] = "/sys/hypervisor/properties/features";
```
Comment (keszybz) [inline src/basic/virt.c]: There's no need to define a variable for this. Just use `#define`.

File: src/basic/virt.c [hunk 2]
```diff
@@ -227,12 +227,31 @@ static int detect_vm_xen(void) {
...
         char *cap, *i;
         int r;
 
+        r = read_one_line_file(path_features, &domcap);
+        if (r < 0 && r != -ENOENT)
+                return r;
+	if (r == 0) {
+                unsigned long features = 0;
```
Comment (keszybz) [inline src/basic/virt.c]: No need to initialize this.
Also please add an empty line here (we add an empty like between variable declarations and code).

Comment (poettering) [reply src/basic/virt.c]: also, weird indentation. please stick to multiples of 8 spaces. See CODING_STYLE

File: src/basic/virt.c [hunk 3]
```diff
@@ -227,12 +227,32 @@ static int detect_vm_xen(void) {
...
         char *cap, *i;
         int r;
 
+        r = read_one_line_file(PATH_FEATURES, &domcap);
+        if (r < 0 && r != -ENOENT)
+                return r;
+	if (r == 0) {
+                unsigned long features;
+
+                r = sscanf(domcap, "%lx", &features);
```
Comment (poettering) [inline src/basic/virt.c]: i think safe_atolu() would be nicer here

Comment (olafhering) [reply src/basic/virt.c]: Thanks, I was looking for helpers like that, but missed them. Then I found many users of sscanf and used it as well.

Comment (keszybz) [reply src/basic/virt.c]: I thought about `safe_atolu`, but this rejects trailing whitespace. I don't have a XEN system around to test if it's present, but I think it'd be safest to explicitly truncate the string using `strstrip`.

**Comment by keszybz** [comment]
Oh, you already pushed, I was late with my comment. If it's works, I guess it's OK without strstrip.

File: src/basic/virt.c (hunk 3 above)
Comment (olafhering) [reply src/basic/virt.c]: At this point the kernel just does `snprintf(buf,"%x", val)` (drivers/xen/sys-hypervisor.c), and I think sscanf would just cope with upcoming changes to the content. The kernel code suggests that at some point the content might be ` %lx%lx `, and this function would need any update either way to handle the adjusted interface. Perhaps this function should assume that only the leading 8 chars should be consumed. That is, the feature string should be truncated at b[8].

But then, if `%lx%lx...` is really a good interface is a separate discussion, which should take place once XENFEAT_NR_SUBMAPS becomes something else than 1.

**Comment by olafhering** [comment]
Not using sscanf() causes a regression. Discussion triggered on systemd-devel about how to parse `snprintf(buf,"%x", val)` input.

**Pull request merged**